conwayste/src/ui/label.rs

```diff
@@ -0,0 +1,134 @@
+/*  Copyright 2019 the Conwayste Developers.
+ *
+ *  This file is part of conwayste.
+ *
+ *  conwayste is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  conwayste is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with conwayste.  If not, see
+ *  <http://www.gnu.org/licenses/>. */
+
+use ggez::{Context, GameResult};
+use ggez::graphics::{self, Color, Rect, Font, Text, TextFragment, DrawParam, Drawable};
+use ggez::nalgebra::{Point2, Vector2};
+
+use std::rc::Rc;
+
+use super::{
+    widget::Widget,
+    WidgetID
+};
+
+use crate::constants::DEFAULT_UI_FONT_SCALE;
+
+pub struct Label {
+    pub id: WidgetID,
+    pub textfrag: TextFragment,
+    pub dimensions: Rect,
+}
+
+/// A graphical widget representation of text
+impl Label {
+    /// Creates a Label widget.
+    ///
+    /// # Arguments
+    /// * `ctx` - GGEZ context
+    /// * `widget_id` - Unique widget identifier
+    /// * `font` - font to be used when drawing the text
+    /// * `string` - Label text
+    /// * `color` - Text color
+    /// * `dest` - Destination point
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use ui::Label;
+    ///
+    /// fn new(ctx: &mut Context) -> GameResult<MainState> {
+    ///     let font = Rc::new(Font::default());
+    ///     let label = Label::new(ctx,
+    ///         WidgetID::TestLabel,
+    ///         font,
+    ///         "TestButton",
+    ///         Color::from(css::DARKCYAN),
+    ///         Color::from(css::WHITE)
+    ///     );
+    ///
+    ///     label.draw(ctx)?;
+    /// }
+    /// ```
+    ///
+    pub fn new(ctx: &mut Context, widget_id: WidgetID, font: Rc<Font>, string: String, color: Color, dest: Point2<f32>) -> Self {
+        let font: Font = *font;
+
+        let text_fragment = TextFragment::new(string.clone())
+            .scale(*DEFAULT_UI_FONT_SCALE)
+            .color(color)
+            .font(font);
+
+        let text = Text::new(text_fragment.clone());
+        // unwrap safe b/c if this fails then the game is fundamentally broken and is not in a usable state
+        let mut dimensions = <Text as Drawable>::dimensions(&text, ctx).unwrap();
```

Comment:
It might be better if the dimensions were supplied by the user and not determined by the text size. If the text is too big, something should be done (like truncation from left or right) to make it fit.

This has to do with the `draw` method too.

Reply:
It looks like ggez already supports wrappable text! https://docs.rs/ggez/0.5.0/ggez/graphics/struct.Text.html#method.set_bounds

Reply:
I think by specifying the dimensions in the function signature we introduce unnecessary complexity. Speaking from personal experience, it becomes a repetitive game of sizing-the-rectangle. If anything, I would accept leaving the current implementation as the default behavior and adding support for wrap-ability via a builder pattern.